File: app/src/main/java/com.tv.lg_webos/MainActivity.java
package com.tv.lg_webos;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    public static boolean m_debugMode = true;
    private LGTV mTv;
    private EditText ipInput;
    
    // Variables for Trackpad Math
    private float lastX, lastY;
    private boolean isMoving = false;

    // Command List for LG TV
    public enum KEY_INDEX {
        TV, YOUTUBE, NETFLIX, AMAZON,
        ON, OFF, MUTE, VOLUME_INCREASE, VOLUME_DECREASE,
        CHANNEL_INCREASE, CHANNEL_DECREASE, 
        HOME, BACK, UP, DOWN, LEFT, RIGHT, ENTER, EXIT
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main); // Loads the native XML

        // 1. Initialize Backend
        mTv = new LGTV(this);
        mTv.loadMainPreferences();

        // 2. Setup IP and Connect Button
        ipInput = findViewById(R.id.ip_input);
        if (mTv.getMyIP() != null) ipInput.setText(mTv.getMyIP());

        findViewById(R.id.btn_connect).setOnClickListener(v -> {
            String ip = ipInput.getText().toString();
            mTv.setMyIP(ip);
            mTv.saveIPPreference();
            mTv.TV_Pairing();
            Toast.makeText(this, "Connecting...", Toast.LENGTH_SHORT).show();
        });

        // 3. Setup Buttons (Simple Helper Function)
        setupBtn(R.id.btn_vol_up, "VOL_UP", KEY_INDEX.VOLUME_INCREASE);
        setupBtn(R.id.btn_vol_down, "VOL_DOWN", KEY_INDEX.VOLUME_DECREASE);
        setupBtn(R.id.btn_ch_up, "CH_UP", KEY_INDEX.CHANNEL_INCREASE);
        setupBtn(R.id.btn_ch_down, "CH_DOWN", KEY_INDEX.CHANNEL_DECREASE);
        
        setupBtn(R.id.btn_up, "UP", KEY_INDEX.UP);
        setupBtn(R.id.btn_down, "DOWN", KEY_INDEX.DOWN);
        setupBtn(R.id.btn_left, "LEFT", KEY_INDEX.LEFT);
        setupBtn(R.id.btn_right, "RIGHT", KEY_INDEX.RIGHT);
        setupBtn(R.id.btn_enter, "ENTER", KEY_INDEX.ENTER);
        
        setupBtn(R.id.btn_back, "BACK", KEY_INDEX.BACK);
        setupBtn(R.id.btn_home, "HOME", KEY_INDEX.HOME);

        // 4. Setup Native Trackpad
        setupTrackpad();
    }

    // Helper to connect a button to a TV command
    private void setupBtn(int id, String logName, KEY_INDEX key) {
        Button btn = findViewById(id);
        if (btn != null) {
            btn.setOnClickListener(v -> mTv.send_key(logName, key));
        }
    }

    // The Trackpad Logic
    private void setupTrackpad() {
        View trackpad = findViewById(R.id.trackpad_view);
        
        trackpad.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        // Finger touched screen - save starting position
                        lastX = event.getX();
                        lastY = event.getY();
                        isMoving = false;
                        return true;

                    case MotionEvent.ACTION_MOVE:
                        // Finger moved - calculate difference (delta)
                        isMoving = true;
                        int dx = (int) (event.getX() - lastX);
                        int dy = (int) (event.getY() - lastY);
                        
                        // Send move command to TV
                        mTv.movePointer(dx, dy);
                        
                        // Update last position for next movement
                        lastX = event.getX();
                        lastY = event.getY();
                        return true;

                    case MotionEvent.ACTION_UP:
                        // Finger lifted - if we didn't move much, treat as a CLICK
                        if (!isMoving) {
                            mTv.clickPointer();
                            v.performClick(); // Accessibility helper
                        }
                        return true;
                }
                return false;
            }
        });
    }
} 

File: app/src/main/java/com.tv.lg_webos/MainActivity.kt
package com.tv.lg_webos

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.HapticFeedbackConstants
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.EditText
import android.widget.Toast
import android.widget.ViewFlipper
import androidx.appcompat.app.AppCompatActivity

// Assuming LGTV.java is still in Java (Interop works perfectly)
class MainActivity : AppCompatActivity() {

    private lateinit var mTv: LGTV
    private lateinit var viewFlipper: ViewFlipper
    private lateinit var ipInput: EditText

    // Touchpad State
    private var lastX = 0f
    private var lastY = 0f
    private var scrollX = 0f
    private var scrollY = 0f
    private var isMoving = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        
        // 1. Immersive Fullscreen Mode
        window.setFlags(
            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
        )
        setContentView(R.layout.activity_main)

        // 2. Initialize Backend
        mTv = LGTV(this)
        mTv.loadMainPreferences()

        setupUI()
    }

    private fun setupUI() {
        // --- Status Bar Spacer ---
        // (Optional: Adjust top padding if content is hidden behind notch)
        
        // --- Header & Connection ---
        viewFlipper = findViewById(R.id.view_flipper)
        ipInput = findViewById(R.id.ip_input)
        
        mTv.myIP?.let { ipInput.setText(it) }

        findViewById<View>(R.id.btn_link).setOnClickListener { view ->
            animateButton(view)
            val ip = ipInput.text.toString()
            mTv.myIP = ip
            mTv.saveIPPreference()
            mTv.TV_Pairing()
            Toast.makeText(this, "Searching for $ip...", Toast.LENGTH_SHORT).show()
        }

        // --- Tabs (Neon Switcher) ---
        setupTab(R.id.tab_nav, 0)
        setupTab(R.id.tab_touch, 1)
        setupTab(R.id.tab_type, 2)
        
        // Set default
        findViewById<View>(R.id.tab_nav).isSelected = true

        // --- Navigation Keys ---
        setupButton(R.id.btn_ok, "ENTER", MainActivity.KEY_INDEX.ENTER)
        setupButton(R.id.btn_up, "UP", MainActivity.KEY_INDEX.UP)
        setupButton(R.id.btn_down, "DOWN", MainActivity.KEY_INDEX.DOWN)
        setupButton(R.id.btn_left, "LEFT", MainActivity.KEY_INDEX.LEFT)
        setupButton(R.id.btn_right, "RIGHT", MainActivity.KEY_INDEX.RIGHT)

        // --- Media & Volume ---
        setupButton(R.id.btn_vol_up, "VOL+", MainActivity.KEY_INDEX.VOLUME_INCREASE)
        setupButton(R.id.btn_vol_down, "VOL-", MainActivity.KEY_INDEX.VOLUME_DECREASE)
        setupButton(R.id.btn_ch_up, "CH+", MainActivity.KEY_INDEX.CHANNEL_INCREASE)
        setupButton(R.id.btn_ch_down, "CH-", MainActivity.KEY_INDEX.CHANNEL_DECREASE)
        setupButton(R.id.btn_mute, "MUTE", MainActivity.KEY_INDEX.MUTE)
        setupButton(R.id.btn_home, "HOME", MainActivity.KEY_INDEX.HOME)
        setupButton(R.id.btn_back, "BACK", MainActivity.KEY_INDEX.BACK)
        setupButton(R.id.btn_power, "POWER", MainActivity.KEY_INDEX.ON)

        // --- App Shortcuts ---
        setupButton(R.id.btn_netflix, "NETFLIX", MainActivity.KEY_INDEX.NETFLIX)
        setupButton(R.id.btn_youtube, "YOUTUBE", MainActivity.KEY_INDEX.YOUTUBE)

        // --- Touchpad Logic ---
        setupTouchpad()
    }

    private fun setupTab(id: Int, childIndex: Int) {
        findViewById<View>(id).setOnClickListener { v ->
            v.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
            
            // Reset all tabs
            findViewById<View>(R.id.tab_nav).isSelected = false
            findViewById<View>(R.id.tab_touch).isSelected = false
            findViewById<View>(R.id.tab_type).isSelected = false
            
            // Select new
            v.isSelected = true
            viewFlipper.displayedChild = childIndex
        }
    }

    private fun setupButton(id: Int, logName: String, key: MainActivity.KEY_INDEX) {
        findViewById<View>(id).setOnClickListener { v ->
            animateButton(v)
            mTv.send_key(logName, key)
        }
    }

    // --- "World Best" Animation Engine ---
    private fun animateButton(v: View) {
        v.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
        v.animate()
            .scaleX(0.9f).scaleY(0.9f)
            .setDuration(50)
            .setInterpolator(AccelerateDecelerateInterpolator())
            .withEndAction {
                v.animate().scaleX(1f).scaleY(1f).setDuration(100).start()
            }
            .start()
    }

    private fun setupTouchpad() {
        val pad = findViewById<View>(R.id.touch_pad_surface) ?: return
        
        pad.setOnTouchListener { v, event ->
            when (event.actionMasked) {
                MotionEvent.ACTION_DOWN -> {
                    lastX = event.x
                    lastY = event.y
                    isMoving = false
                    true
                }
                MotionEvent.ACTION_POINTER_DOWN -> {
                    if (event.pointerCount == 2) {
                        scrollX = event.getX(0)
                        scrollY = event.getY(0)
                    }
                    true
                }
                MotionEvent.ACTION_MOVE -> {
                    isMoving = true
                    if (event.pointerCount == 1) {
                        val dx = (event.x - lastX).toInt()
                        val dy = (event.y - lastY).toInt()
                        mTv.movePointer(dx, dy)
                        lastX = event.x
                        lastY = event.y
                    } else if (event.pointerCount == 2) {
                        val dx = ((event.getX(0) - scrollX) / 2).toInt()
                        val dy = ((event.getY(0) - scrollY) / 2).toInt()
                        mTv.scroll(dx, dy)
                        scrollX = event.getX(0)
                        scrollY = event.getY(0)
                    }
                    true
                }
                MotionEvent.ACTION_UP -> {
                    if (!isMoving) {
                        v.performHapticFeedback(HapticFeedbackConstants.CONTEXT_CLICK)
                        mTv.clickPointer()
                        animateButton(v)
                    }
                    true
                }
                else -> false
            }
        }
    }
}
